README.md
# Unidimensional and Multidimensional Methods for Recurrence Quantification Analysis with crqa.

The crqa, R package, allows users to conduct a wide range of recurrence-based analyses on single (e.g., auto-recurrence) and multivariate time series (e.g., multidimensional cross-recurrence quantification), examine coupling properties underlying leader-follower relationships (i.e., diagonal-profile methods), as well as, track the evolution of recurrence rate over the time course (i.e., windowed methods). 

## Installation

``` r
# You can install the latest version of crqa on CRAN by running:
install.packages("crqa")

# Or for the development version from GitHub:
# install.packages("devtools")
devtools::install_github("morenococo/crqa")
```

# Usage

crqa comes with some data that can be used to test and study the different functions therein. 

``` r
data(crqa) # load the data 
```

## RQA on a categorical time-series (auto-recurrence)

First, specify the arguments that will be used in the crqa, core function.

``` r
## parameter setting 
delay = 1; embed = 1; rescale = 0; radius = 0.0001;
normalize = 0; mindiagline = 2; minvertline = 2;
tw = 1; whiteline = FALSE; recpt = FALSE; 
side = "both"; method = 'rqa'; metric = 'euclidean';  
datatype = "categorical"
```

Then, run crqa on a nursery rhyme “The wheels on the bus” by Verna Hills: a vector of 120 strings (i.e., the words of the song),

``` r
ans = crqa(text, text, delay, embed, rescale, radius, normalize, 
           mindiagline, minvertline, tw, whiteline, recpt, side, method, metric, 
           datatype)
```

Have a look at the output, which contains different measures extracted from the recurrence plot (RP), and the RP itself, which can be plotted using the plotRP function. 

``` r
str(ans)
```
## CRQA on a categorical time-series (cross-recurrence)

Cross-recurrence extends univariate recurrence analysis into a bivariate analysis that allows quantification of the temporal coupling properties of two time series. We use eye-tracking data, 2,000 observations of six possible screen locations that are looked at by a dyad engaged into a joint task.  

``` r
listener = eyemovement$listener
narrator = eyemovement$narrator

delay = 1; embed = 1; rescale = 0; radius = .01;
normalize = 0; mindiagline = 2; minvertline = 2;
tw = 0; whiteline = FALSE; recpt = FALSE; side = "both"
method = 'crqa'; metric = 'euclidean';  
datatype = "categorical"

ans = crqa(narrator, listener, delay, embed, rescale, radius, normalize, 
           mindiagline, minvertline, tw, whiteline, recpt, side, method, metric, 
           datatype)
```

### Diagonal cross-recurrence profile. 

From cross-recurrence plots is possible to extract the diagonal cross-recurrence profiles (DCRPs) and use them to capture leader-follower-relationships. 

``` r
timecourse = round( seq(-3300,3300,33)/1000, digit = 2)  ## construct the time-course for the diagonal profile

res = drpfromts(narrator, listener, windowsize = 100,
                 radius = 0.001, delay = 1, embed = 1, rescale = 0,
                 normalize = 0, mindiagline = 2, minvertline = 2,
                 tw = 0, whiteline = F, recpt = F, side = 'both', 
                 method = 'crqa', metric = 'euclidean', 
                 datatype = 'continuous')
                 
 ## visualise the diagonal-profile
 profile = res$profile*100 ## extract it from the res object
 
plot(timecourse, profile, type = "l", lwd = 2.5, xlab = "Lag (seconds)",
     ylab = "Recurrence Rate %")                 
```

## Multidimensional cross-recurrence quantification analysis

Multidimensional cross-recurrence quantification analysis allows for the computation of cross-recurrences between two multidimensional time-series. We use hand-movement data from a complex LEGO joint construction task. The dataframe comprises of 5,799 observations.

``` r
# reduce the dimensionality of the time series to make the computation faster
# handset = handmovement[1:3000, ]
handset = handmovement[1:1000, ]

P1 = cbind(handset$P1_TT_d, handset$P1_TT_n) 
P2 = cbind(handset$P2_TT_d, handset$P2_TT_n)

delay = 5; embed = 2; rescale = 0; radius = .1;
normalize = 0; mindiagline = 10; minvertline = 10;
tw = 0; whiteline = FALSE; recpt = FALSE; side = "both"
method = 'mdcrqa'; metric = 'euclidean';  
datatype = "continuous"

ans = crqa(P1, P2, delay, embed, rescale, radius, normalize, 
           mindiagline, minvertline, tw, whiteline, recpt, side, method, metric, 
           datatype)

RP = ans$RP
results = unlist(ans[1:10])
print(results)
```

## Authors

* **Moreno I Coco** - *role = [cre, aut]* - (moreno.cocoi@gmail.com)
* **Dan Mønster** - *role = [aut]* - (danm@econ.au.dk)
* **Giuseppe Leonardi** - *role = [aut]* - (g.leonardi@vizja.pl)
* **Rick Dale** - *role = [aut]* - (rdale@ucla.edu)
* **Sebastian Wallot** - *role = [aut] - (sebastian.wallot@ae.mpg.de)

## Acknowledgments

* **James D. Dixon** - *role = [ctb]* - (james.dixon@uconn.edu)
* **John  C. Nash** - *role = [ctb]* -  (nashjc@uottawa.ca)

NEWS.md
# crqa (cross-recurrence quantification analysis)

# crqa 2.0.3

* Fixed minor bug on drpfromts() when method was `mdcrqa`. The argument `datatype` was forcing the input data to be in vector form, and should instead be left as a matrix. Added if statement to check (line 25)

# crqa 2.0.1

* Fixed bug on wincrqa() and windowdrp() to run windowed RQA on multidimensional data.
    * Simplified output of wincrqa() now returning a dataframe
    * Included new import from FSA() package to use diags() function. 

# crqa 2.0.0

# Major features

* Extension of recurrence analysis to multidimensional time-series data, and significant update of computational procedures in `crqa()`
    * Added arguments: `method`, `metric` and `datatype`
    * Improved method for phase-space reconstruction (line 207:237)
    * Included computation of categorical entropy (line 411:420)
    * Removal of argument `checkl`
    
* Simplified structure of functions and better division between core and ancillary functions:
    * `crqa_helpers` contains several functions previously exported (e.g., `theiler` or `tt`) that are now only accessed internally by the `crqa()` package.
    
* New functions `mdDelay()` and `mdFnn()` to estimate Average Mutual Information and False Nearest Neighbours of multidimensional time-series.

* Experimental `piecewiseRQA()` function created to better handle the computational load of large time-series.

* `optimizeParam()` now works also with multidimensional time-series. 
    
# Minor features

* Deprecated functions: `CTcrqa()`, `runcrqa()`, `calcphi()`, `takephi()`

* On `crqa()` 
    * improved error checking and warning messages (line 133:170)
    * added two more ways of rescaling the distance matrix (line 252:272)
    
* On `plotRP()` added arguments to improve the plotting of Recurrence Plots 

* `drpdfromts()` is now called `drpfromts()` and it has been rewritten to align with the new version of `crqa()`.

* A convenience function called `numerify` in `crqa_helpers` is automatically called when a user inputs categorical series (i.e., it contains either characters or factors) and this function is used to recode the levels of such time-series into numerical codes (to run crqa). A warning is send to the user when `crqa()`.

* `windowdrp()` has been rewritten to align with the new version of `crqa()`.  

* `wincrqa()` has been rewritten to align with the new version of `crqa()` and better names for the output were provided.  

# crqa 1.0.9

* On `drpdfromts()` removed a left over constant used for testing (line: 42)

# crqa 1.0.8

## Improvements and bug fixes

* On `drpdfromts()` fixed initialisation of dimensions for empty RP (line: 52)

# crqa 1.0.7

## New functions

* `plotRP()` convenience function based on the standard `plot()` to visualize a Recurrence Plot

## Improvements and bug fixes

* On `crqa()` added a few more checks (`stop`) if the data inputted did not comply with the function, and send a warning message.

* `drpdfromts()` entirely rewritten around `crqa()` to better deal with continuous valued time-series.

* `runcrqa()` fixed to fit with the revised functions: `drpdfromts()` and `windowdrp()`, 

* `tt()` fixed `rBind` (line 23 and 91) which was deprecated from the `Matrix()` package.

* `windowdrp()`entirely rewritten around the new version of `drpdfromts()` 

* `wincrqa()` adjusted indexing of windows (line: 40:41)

# crqa 1.0.6

## New functions

* `ami()` externalised from `optimizeParam()`

* `lorenzattractor()` simulates and plots 3D data from a Lorenz Attractor

## Improvements and bug fixes

* On `crqa()` include a `stop` (line: 95:100) if time-series were shorter than their phase space reconstructed portaits

* On `optimizeParam()` included argument `typeami` to set the type of `ami()` desired (either, minimum dip or maximum lag)

# crqa 1.0.5

## Improvements and bug fixes

* On `checkts()`. added argument `pad` (line: 38:67), which gives the option, in case of series of different length to extend the shortest sequence either with mean value (if the variable is in a continuous scale) or with a random label not present in either series (if the variable is categorical).

* On `crqa()`: 
    * Added default values for arguments in the call of the function. 
    * Included the argument `side` to select the region of the Recurrence Plot to extract measures on
    * Included the argument `checkl`, a wrapper to call the function `checkts()` directly inside this function. 

* On `optimizeParam()`:
    * Added arguments `min.rec` and `max.rec` to the call. 
    * Added argument within `par` (`fnnpercent`) to estimate False Nearest Neighbours based on a percentage reduction with respect to first dimension (line: 199:241)
    * Simplified code throughout and markedly improved the estimation of the radius (line: 299:360)
    
* On `runcrqa()` included argument `pad` in the call to work with the revised version of `checkts()`

* On `wincrqa()` added calculation of TREND (line: 64:87) 

# crqa 1.0.4

* crqa 1.0.3 did not pass CRAN check because of `DESCRIPTION` (`Depends` field) and was resubmitted as 1.0.4.

# crqa 1.0.3

## New functionx 

* Added `theiler()` function to choose the separation between values on the time series when specifying a delay reconstruction vector, i.e., the Theiler window.

## Improvements and bug fixes

* On `crqa()` added theiler window (`theiler()`, line: 170:175)

* On `optimizeParam()` improved calculation of average mutual information (line 56:119), and added estimation of radius within user specified expected recurrence values (line 237:284)

* On `runcrqa()` added missing arguments when calling `wincrqa()` (line 104-110:112) 

* On `tt()` simplified calculation of laminarity (line 65)

* On `wincrqa()` added missing arguments in the function call to exploit better functionality in `crqa()`

# crqa 1.0.2

R/wincrqa.R
## written by Moreno I. Coco 2013 (moreno.cocoi@gmail.com)
## original Matlab code by Rick Dale

## calculate recurrence over different sized windows
## arguments: step = interval considered on the serie;
##            windowstep = the step of the window. 
##            windowsize =  the size of the window
##           
##          
## other arguments to pass are the same as crqa

# windowsize = 200; windowstep = 50;
# type = 1; delay = 1; embed = 1;
# rescale = 1; radius = 0.0001;
# normalize = 0; minline = 2

# source("crqa.R")

# tS = simts(0.25, 0.05, 0.2, 0.2, 0.25, 1000)
# ts1 = tS[1,]; ts2 = tS[2,]

.packageName <- 'crqa'

wincrqa <- function(ts1, ts2, windowstep, windowsize, delay, embed,
                    radius = 0.001, rescale = 0,
                    normalize = 0, mindiagline = 2, minvertline = 2,
                    tw = 0, whiteline = FALSE, recpt = FALSE, side = 'both', 
                    method = 'crqa', metric = 'euclidean', 
                    datatype = 'continuous', trend = FALSE){
  
  irregular = FALSE; # initialize a flag that will be used to check whether the last window is included
  ## check the different contexts in which the analyses are running
  if (method == "rqa"){
    ## check that the user as provided the same data to allow windowed recurrence
    if (exists("ts2")) ts2 = ts2 else stop("Please provide the same vector as argument of t2")
  }
  
  if (method == "crqa"){
    ts1 = as.vector(as.matrix(ts1));   ts2 = as.vector(as.matrix(ts2))
    maxd = length(ts1) ## the total number of points
    points = seq(1, (maxd - (windowsize)-1), windowstep)
  }
  
  if (method == "mdcrqa"){
    ts1 = as.matrix(ts1);   ts2 = as.matrix(ts2)
    maxd = nrow(ts1) ## the total number of points
    points = seq(1, (maxd - (windowsize)-1), windowstep)
  }
  
  ## need to make sure to include also "irregular" final segments
  fpoint = points[length(points)] ## beginning of last window
    ## as the dimension of the data often may not be that precise
    if (fpoint != (fpoint + windowsize)){
      irregular = TRUE ## flag used later on
      fpoint = fpoint + windowsize # update last point
      points = c(points,  fpoint) ## add final point
  }
  
  crqwin = vector()
  tsp = 0 ## set a counter with the win at which rec was found    
  
  # i = 1621
  for (i in points){
    tsp = tsp +1
    
    if (irregular == TRUE & i == fpoint){ ## this is the last window to consider 
      ixs = i:maxd 
      # warning(paste("Your latest window was shorter: ", length(ixs)))
    } else {
    ixs = i:(i + windowsize - 1)
    }
    
    if (method == "crqa" | method == "rqa"){
      ## the data consists of two vectors  
      ts1win = ts1[ixs];
      ts2win = ts2[ixs];
    }
    
    if (method == "mdcrqa"){
      ## the data consists of two matrices 
      ts1win = ts1[ixs, ];
      ts2win = ts2[ixs, ];
    }
    
    ans = crqa(ts1win, ts2win, delay, embed, rescale,
               radius, normalize, mindiagline, minvertline,
               tw, whiteline, recpt, side, method, metric,
               datatype)
    
    RP = ans$RP
    if (length(RP) == 1) RP = vector() ## a trick for cases
    ## with empty recurrence plot
    ans = as.numeric( unlist(ans[1:10]) )
    ## if trend needs to be calculated do it here
    
    if (trend == TRUE){
      
      if (length(RP) > 0){
        RP = as.matrix(RP) ## diags() has changed behaviour 
        NX = ncol(RP)
        T = vector("numeric", length = NX-1)
    
  
        ## below some test matrices
        # RP = matrix(1:16,nrow=4)
        # RP = cbind(c(2,7), c(1,9), c(10, 12))
        
        # do we need to span the full matrix or just half?
        ## create a vector of indeces for all diagonals of the matrix  
        # ixi_diag = -(ncol(RP)-1):(nrow(RP)-1) 
        
        # k = 1
        for (k in 1:(NX-1)){  ##ixi_diag
#          print(diags(RP, k)) ## check the diagonal
          ixi_rec = which(diags(RP, k) != F) ## find recurrence
          T[k] = length(ixi_rec)/ (NX-k)*100;
        }
        
        Ntau = NX - 1 - round(0.1*NX);
        
        ## last 10% of the RP will be skipped
        p = polyfit(2:(Ntau+1),T[1:Ntau], 1) # slope
        TREND = 1000 * p[1]
        ## Webber's definition includes factor 1000
        
      } else { TREND = NA}
    } else {
      TREND = NA
    }
    
    crqwin = rbind(crqwin, c(ans, tsp, TREND), deparse.level = 0)
    
  }
  
  ## name the measures
  colnames(crqwin) = c("RR", "DET", "NRLINE", "maxL", "L", 
                       "ENTR", "rENTR", "LAM", "TT", "catH", "win", "TREND")
  
  return(as.data.frame(crqwin))
  
}

